src/timeout_wheel.rs

```diff
@@ -0,0 +1,41 @@
+use std::collections::{BTreeMap, HashMap};
+use std::hash::Hash;
+
+use chrono::{DateTime, Utc};
+
+#[derive(Default)]
+pub(crate) struct TimeoutWheel<T: Clone + Hash + std::cmp::Eq> {
+    wheel: BTreeMap<DateTime<Utc>, T>,
+    map: HashMap<T, DateTime<Utc>>,
+}
+
+impl<T: Clone + Hash + std::cmp::Eq> TimeoutWheel<T> {
+    pub fn new() -> Self {
+        TimeoutWheel {
+            wheel: BTreeMap::new(),
+            map: HashMap::new(),
+        }
+    }
+
+    pub fn insert(&mut self, e: T, timeout: DateTime<Utc>) {
+        self.wheel.insert(timeout, e.clone());
+        self.map.insert(e, timeout);
+    }
+
+    pub fn pop_expired(&mut self) -> Option<T> {
+        self.wheel.first_entry().and_then(|entry| {
+            if entry.key() < &Utc::now() {
```

Comment:
If you hate the next person to read that code, you can use [`Option::filter`](https://doc.rust-lang.org/std/option/enum.Option.html#method.filter) instead of `if`